### PR TITLE
Analyser for carpool parkings in France

### PR DESCRIPTION
**Pull request by PanierAvide**
Hello,

I would like to add this analyser, which enable conflation with unified carpool parking dataset ([BNCLC](https://www.data.gouv.fr/fr/datasets/base-nationale-consolidee-des-lieux-de-covoiturage/)). Everything might be set correctly, but I still run into an issue when launching code on Docker :

```
2019-10-09 14:48:17   error: error on analyse merge_carpool_FR...
2019-10-09 14:48:17     Traceback (most recent call last):
2019-10-09 14:48:17       File "./osmose_run.py", line 230, in execc
2019-10-09 14:48:17         analyser_obj.analyser()
2019-10-09 14:48:17       File "/opt/osmose-backend/analysers/Analyser_Osmosis.py", line 192, in analyser
2019-10-09 14:48:17         self.analyser_osmosis_common()
2019-10-09 14:48:17       File "/opt/osmose-backend/analysers/Analyser_Merge.py", line 867, in analyser_osmosis_common
2019-10-09 14:48:17         table = self.load.run(self, self.parser, self.mapping, self.config.db_user, self.__class__.__name__.lower()[15:], self.analyser_version())
2019-10-09 14:48:17       File "/opt/osmose-backend/analysers/Analyser_Merge.py", line 631, in run
2019-10-09 14:48:17         header = parser.header()
2019-10-09 14:48:17       File "/opt/osmose-backend/analysers/Analyser_Merge.py", line 413, in header
2019-10-09 14:48:17         return csv.reader(self.f, delimiter=self.separator, quotechar=self.quote).next()
2019-10-09 14:48:17       File "/usr/local/lib/python3.7/site-packages/backports/csv.py", line 415, in __next__
2019-10-09 14:48:17         self.parse_process_char(c)
2019-10-09 14:48:17       File "/usr/local/lib/python3.7/site-packages/backports/csv.py", line 269, in parse_process_char
2019-10-09 14:48:17         return switch[self.state](c)
2019-10-09 14:48:17       File "/usr/local/lib/python3.7/site-packages/backports/csv.py", line 382, in _parse_eat_crnl
2019-10-09 14:48:17         raise Error('new-line character seen in unquoted field - do you '
2019-10-09 14:48:17     _csv.Error: new-line character seen in unquoted field - do you need to open the file in universal-newline mode?
```

It might be related to the input file having some line return in it in quoted fields, but I don't know if any option is existing in Osmose to handle this properly.

Regards,

Adrien.

### CODE REVIEW

**Comment by frodrigo** [comment]
It should be OK. But some one already report issue with this. I don't know about it.
Maybe some parameters are missing:

https://github.com/osm-fr/osmose-backend/blob/master/analysers/Analyser_Merge.py#L413

**Comment by frodrigo** [comment]
Yes 'rU' (no 'rb').

File: analysers/analyser_merge_carpool_FR.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+                generate = Generate(
+                    static1 = {"amenity": "parking", "carpool": "designated"},
+                    static2 = {"source": self.source},
+                    mapping1 = {
+                        "ref:FR:BNCLC": "id_lieu",
+                        "name": "nom_lieu",
+                        "capacity": "nbre_pl",
+                        "capacity:disabled": "nbre_pmr",
+                        "lit": lambda res: "yes" if res["lumiere"] == "true" else ("no" if res["lumiere"] == false else None)},
+                    text = lambda tags, fields: {"en": u"Carpool parking %s" % fields[u"nom_lieu"]} )))
```
Comment (frodrigo) [inline analysers/analyser_merge_carpool_FR.py]:
```suggest
                   text = lambda tags, fields: T_f(u"Carpool parking {0}", fields[u"nom_lieu"]) )))
```

File: analysers/analyser_merge_carpool_FR.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+###########################################################################
+
+from .Analyser_Merge import Analyser_Merge, Source, CSV, Load, Mapping, Select, Generate
+
+
+class Analyser_Merge_Carpool_FR(Analyser_Merge):
+    def __init__(self, config, logger = None):
+        self.missing_official = {"item":"8130", "class": 31, "level": 3, "tag": ["merge", "parking", "carpool"], "desc": T_(u"Carpool parking not integrated") }
+        self.possible_merge   = {"item":"8131", "class": 33, "level": 3, "tag": ["merge", "parking", "carpool"], "desc": T_(u"Carpool parking integration suggestion") }
+        self.update_official  = {"item":"8132", "class": 34, "level": 3, "tag": ["merge", "parking", "carpool"], "desc": T_(u"Carpool parking update") }
```
Comment (frodrigo) [inline analysers/analyser_merge_carpool_FR.py]: Change to class 41, 43, 44.

**Comment by PanierAvide** [comment]
I finally managed to make this work. So I added an option in `Source` class for managing this universal new line option. There was also an encoding issue with the CSV file, so I switched to "utf-8-sig". Results are now correctly generated.

**Comment by PanierAvide** [comment]
Well the error is happening due to the way file is opened. I think it only happens on formats relying on line breaks (most common being CSV). But as it is happening due to how file is opened, I'm not sure this can go in CSV class, as it is only handling parsing once text is read.

**Comment by frodrigo** [comment]
But we reopen the file here https://github.com/osm-fr/osmose-backend/blob/master/analysers/Analyser_Merge.py#L411

**Comment by PanierAvide** [comment]
OK, so I moved the option universalNewLine to `CSV`, and option is passed instead when calling `open` method in `Source` class.

**Comment by frodrigo** [comment]
I an not OK with the changes of Analyser_Merge.py . I will try to fix it.

**Comment by PanierAvide** [comment]
Thanks, because I wasn't expecting to run into this kind of encoding/opening issue while just adding a CSV file. Apparently that was a trap CSV :sweat_smile:

File: analysers/analyser_merge_carpool_FR.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+class Analyser_Merge_Carpool_FR(Analyser_Merge):
+    def __init__(self, config, logger = None):
+        self.missing_official = {"item":"8130", "class": 41, "level": 3, "tag": ["merge", "parking", "carpool"], "desc": T_(u"Carpool parking not integrated") }
+        self.possible_merge   = {"item":"8131", "class": 43, "level": 3, "tag": ["merge", "parking", "carpool"], "desc": T_(u"Carpool parking integration suggestion") }
+        self.update_official  = {"item":"8132", "class": 44, "level": 3, "tag": ["merge", "parking", "carpool"], "desc": T_(u"Carpool parking update") }
+        Analyser_Merge.__init__(self, config, logger,
+            u"https://www.data.gouv.fr/fr/datasets/base-nationale-consolidee-des-lieux-de-covoiturage",
+            u"Base nationale consolidée des lieux de covoiturage",
+            CSV(Source(attribution = u"Transport.data.gouv.fr", millesime = "09/2019", encoding = "utf-8-sig",
+                    fileUrl = u"https://www.data.gouv.fr/fr/datasets/r/e0962ca4-2fb9-4257-a569-56704df3243d"), separator = u";", universalNewLine = True),
```
Comment (frodrigo) [inline analysers/analyser_merge_carpool_FR.py]:
```suggestion
                    fileUrl = u"https://www.data.gouv.fr/fr/datasets/r/e0962ca4-2fb9-4257-a569-56704df3243d",
                    filter = lambda text: text.replace('\r', '\n')), separator = u";"),
```

The issue is in the CSV file, not in the reader.

**Comment by frodrigo** [comment]
https://www.data.gouv.fr/fr/datasets/base-nationale-consolidee-des-lieux-de-covoiturage/#discussion-5da1efe86f4441060e7e8fee

**Comment by frodrigo** [comment]
Please can you remove the chabges in Analyser_Merge.py, thank.

**Comment by PanierAvide** [comment]
Thanks for investigating on this, I uploaded requested changes :wink:

**Comment by frodrigo** [comment]
Merged, thank you.

**Comment by frodrigo** [comment]
http://osmose.openstreetmap.fr/fr/errors/?item=8130,8131,8132&class=41,42,43

**Pull request merged**